sql/analyzer/hoist_select_exists.go

```diff
@@ -159,12 +196,75 @@ func decorrelateOuterCols(e *plan.Subquery, scopeLen int) *hoistSubquery {
 	})
 
 	if same || len(outerFilters) == 0 {
-		return nil
+		return nil, nil
+	}
+
+	nodeAliases, err := getTableAliases(n, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	conflicts, nonConflicted := outsideAliases.findConflicts(nodeAliases)
+	for _, goodAlias := range nonConflicted {
+		target, ok := nodeAliases[goodAlias]
+		if !ok {
+			return nil, fmt.Errorf("node alias %s is not in nodeAliases", goodAlias)
+		}
+		nameable := fakeNameable{name: goodAlias}
+		err = outsideAliases.add(nameable, target)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if len(conflicts) > 0 {
+		for _, conflict := range conflicts {
+
+			// conflict, need to rename
+			newAlias := aliasCreator(conflict)
+			var tree transform.TreeIdentity
+			n, tree, err = renameAliases(n, conflict, newAlias)
+			if err != nil {
+				return nil, err
+			}
+			if tree == transform.SameTree {
+				return nil, fmt.Errorf("tree is unchanged after attempted rename")
+			}
+
+			// rename the aliases in the expressions
+			innerFilters, err = renameAliasesInExpressions(innerFilters, conflict, newAlias)
+			if err != nil {
+				return nil, err
+			}
+			outerFilters, err = renameAliasesInExpressions(outerFilters, conflict, newAlias)
+			if err != nil {
+				return nil, err
+			}
+
+			// alias was renamed, need to get the renamed target before adding to the outside aliases collection
+			nodeAliases, err = getTableAliases(n, nil)
```

Comment:
We need rules to do as little work as possible. If you aren't convinced that this is organized in a way that is simple and verifiable, then i think a better solution is adding more tests that poke holes in correctness, or restructuring into helper functions that can be more directly unit tested. We can't run semantic validation on every intermediate data structure and node we build. If we did this with all of our transformation rules, fast queries like point selects would be 2-3x slower.

Reply:
We want to cache this information once in a format that is re-used throughout analysis (memo), and structured in a centralized way where the specialized components are easier to unit test. There are also other ways to test individual rule input/outputs aggressively, or fuzz test inputs to increase coverage.

Reply:
This portion has been rewritten.